src/supplemental/quic/msquic_dial.c

```diff
@@ -0,0 +1,1256 @@
+			log_warn("reconnect failed due to no resumption ticket.\n");
+			quic_strm_fini(qstrm);
+			nng_free(qstrm, sizeof(quic_strm_t));
+		}
+		*/
+
+		break;
+	case QUIC_CONNECTION_EVENT_RESUMPTION_TICKET_RECEIVED:
+		// A resumption ticket (also called New Session Ticket or NST)
+		// was received from the server.
+		log_warn("[conn][%p] Resumption ticket received (%u bytes):\n",
+		    Connection, Event->RESUMPTION_TICKET_RECEIVED.ResumptionTicketLength);
+		if (d->enable_0rtt == false) {
+			log_warn("[conn][%p] Ignore ticket due to turn off the 0RTT");
+			break;
+		}
+		d->rticket_sz = Event->RESUMPTION_TICKET_RECEIVED.ResumptionTicketLength;
+		if (d->rticket_sz > 4096) {
+			log_error("rticket is too long.");
+			break;
+		}
+		memcpy(d->rticket, Event->RESUMPTION_TICKET_RECEIVED.ResumptionTicket,
+		        Event->RESUMPTION_TICKET_RECEIVED.ResumptionTicketLength);
+		break;
+	case QUIC_CONNECTION_EVENT_PEER_CERTIFICATE_RECEIVED:
+		log_info("QUIC_CONNECTION_EVENT_PEER_CERTIFICATE_RECEIVED");
+
+		// TODO Using mbedtls APIs to verify
+		/*
+		 * TODO
+		 * Does MsQuic ensure the connected event will happen after
+		 * peer_certificate_received event.?
+		 */
+		/*
+		if (QUIC_FAILED(rv = verify_peer_cert_tls(
+				Event->PEER_CERTIFICATE_RECEIVED.Certificate,
+				Event->PEER_CERTIFICATE_RECEIVED.Chain, d->cacert))) {
+			log_error("[conn][%p] Invalid certificate file received from the peer");
+			return rv;
+		}
+		*/
+		break;
+	case QUIC_CONNECTION_EVENT_DATAGRAM_STATE_CHANGED:
+		log_info("QUIC_CONNECTION_EVENT_DATAGRAM_STATE_CHANGED");
+		break;
+	case QUIC_CONNECTION_EVENT_STREAMS_AVAILABLE:
+		log_info("QUIC_CONNECTION_EVENT_STREAMS_AVAILABLE");
+		break;
+	case QUIC_CONNECTION_EVENT_IDEAL_PROCESSOR_CHANGED:
+		log_info("QUIC_CONNECTION_EVENT_IDEAL_PROCESSOR_CHANGED");
+		break;
+	default:
+		log_warn("Unknown event type %d!", Event->Type);
+		break;
+	}
+	return QUIC_STATUS_SUCCESS;
+}
+
+// The clients's callback for stream events from MsQuic.
+// New recv cb of quic transport
+_IRQL_requires_max_(DISPATCH_LEVEL)
+_Function_class_(QUIC_STREAM_CALLBACK) QUIC_STATUS QUIC_API
+msquic_strm_cb(_In_ HQUIC stream, _In_opt_ void *Context,
+	_Inout_ QUIC_STREAM_EVENT *Event)
+{
+	nni_quic_conn *c = Context;
+	nni_aio       *aio;
+	nni_iov *      aiov;
+	unsigned       naiov;
+	uint32_t       rlen, rlen2, rpos;
+	uint8_t       *rbuf;
+	uint32_t       count;
+
+	log_debug("quic_strm_cb triggered! %d", Event->Type);
+	switch (Event->Type) {
+	case QUIC_STREAM_EVENT_SEND_COMPLETE:
+		log_debug("QUIC_STREAM_EVENT_SEND_COMPLETE!");
+		if (Event->SEND_COMPLETE.Canceled) {
+			log_warn("[strm][%p] Data sent Canceled: %d",
+			    stream, Event->SEND_COMPLETE.Canceled);
+		}
+		// Priority msg send
+		if ((aio = Event->SEND_COMPLETE.ClientContext) != NULL) {
+			Event->SEND_COMPLETE.ClientContext = NULL;
+			nni_aio_finish(aio, 0, nni_aio_count(aio));
+			break;
+		}
+		// Ordinary send
+		quic_cb(QUIC_STREAM_EVENT_SEND_COMPLETE, c);
```

Comment:
not sure why do you use embedded cb here, plz make it inline

Reply:
Split msquic and nng

Reply:
And easy to read